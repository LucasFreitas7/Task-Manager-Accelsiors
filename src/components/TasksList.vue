<template>
    <v-container>
      <v-tabs>
          <v-tab @click="handleClickOnFilter('all')">All tasks</v-tab>
          <v-tab @click="handleClickOnFilter('completed')">Done tasks</v-tab>
          <v-tab @click="handleClickOnFilter('active')">Undone Tasks</v-tab>
      </v-tabs>
      <div class="warning-text">
          <p class="text-task" v-if="activeClass === 'completed' && this.allFiltered.length === 0"> You dont have completed tasks</p>
          <p class="text-task" v-if="activeClass === 'all' && this.allFiltered.length === 0"> You dont have any tasks</p>
          <p class="text-task" v-if="activeClass === 'active' && this.allFiltered.length === 0"> You have already completed all your tasks</p>
      </div>
      <v-expansion-panels focusable>
        <v-expansion-panel
          v-for="(task,i) in allFiltered"
          :key="i"
        >
          <v-expansion-panel-header>{{ task.activty}}</v-expansion-panel-header>
          <v-expansion-panel-content>
            <div class="panel-icons">
              <v-icon>schedule</v-icon> = {{durationFormat(task.duration) }}
            </div>
            <div v-if="task.comment != ''" class="panel-icons">
              <v-icon>comment</v-icon> = {{ task.comment }}
            </div>
            <div class="panel-icons">
              <v-icon>calendar_month</v-icon> = {{ task.date }}
            </div>
            <div class="panel-buttons">
              <div>
                <v-btn
                    v-if="task.done === false"
                    class="ma-2"
                    color="success"
                    @click="toogleFetch(task)"
                >
                  Done Task
                </v-btn>
                <v-btn
                    v-else
                    class="ma-2"
                    color="error"
                    @click="toogleFetch(task)"
                  >
                  Undone Task
                </v-btn>
              </div>
              <div v-if="activeClass === 'all'">
                <v-btn
                  class="mx-2"
                  fab
                  dark
                  small
                  color="cyan"
                  @click="emitTaskEdit(i)"
                >
                  <v-icon dark>
                    mdi-pencil
                  </v-icon>
              </v-btn>
              <v-btn
                  class="mx-2"
                  fab
                  dark
                  @click="deleteTask(task)"
                  small
                  color="red"
                >
                  <v-icon dark>
                    delete
                  </v-icon>
              </v-btn>
              </div>
            </div>
          </v-expansion-panel-content>
        </v-expansion-panel>
      </v-expansion-panels>
      <div class="button-delete-done">
        <v-btn
        v-if="activeClass === 'completed' && this.allFiltered.length > 0"
        @click="clearDone"
        tile
        color="error"
      >
        <v-icon left>
          delete
        </v-icon>
        Delete all done tasks
      </v-btn>
      </div>
    </v-container>
  </template>
  
  <script>
  import { mapGetters, mapActions } from "vuex";
    export default {
      name: 'TaskList',
      data: () => ({
        activeClass: 'all',
        allFiltered: []
        }),
      computed:{
        ...mapGetters(['getAll', 'undoneAll', 'doneAll'])
      },
      watch:{
        getAll(newValue){
        this.allFiltered = newValue
        },
        doneAll(newValue, oldValue){
          if(newValue.length < oldValue.length && this.activeClass === 'completed'){
            this.allFiltered = newValue
          }
        },
        undoneAll(newValue, oldValue){
          if(newValue.length < oldValue.length && this.activeClass === 'active'){
            this.allFiltered = newValue
          }
        }
      },
      methods:{
        ...mapActions(['toggleChange', 'deleteTask', 'clearDone']),
        handleClickOnFilter(filter){
            this.activeClass = filter;
            switch (filter) {
            case 'all':
                this.allFiltered = this.getAll;
                break;
            case 'completed':
                this.allFiltered = this.doneAll;
                break;
            case 'active':
                this.allFiltered = this.undoneAll;
                break;
            default:
                break;
        }
      },
      emitTaskEdit(id){
            this.$emit('edit', id)

        },
        toogleFetch(task){
          this.toggleChange(task)
        },
        durationFormat(duration){
          if(duration == 1){
            return '1 hour'
          }
          else if(duration == 0.5){
            return '30 minutes'
          }
          else{
            return duration + 'hours'
          }
        },
      },
      created(){
        this.allFiltered = this.getAll
      }
    }
  </script>
  
  <style>
  .filter-activity{
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    padding: 10px;
  }
  
  .activity-text{
    padding-right: 10px;
  }
  
  .material-icons {
    font-family: 'Material Icons';
    font-weight: normal;
    font-style: normal;
    font-size: 24px;  /* Preferred icon size */
    display: inline-block;
    line-height: 1;
    text-transform: none;
    letter-spacing: normal;
    word-wrap: normal;
    white-space: nowrap;
    direction: ltr;
  
    /* Support for all WebKit browsers. */
    -webkit-font-smoothing: antialiased;
    /* Support for Safari and Chrome. */
    text-rendering: optimizeLegibility;
  
    /* Support for Firefox. */
    -moz-osx-font-smoothing: grayscale;
  
    /* Support for IE. */
    font-feature-settings: 'liga';
  }
  
  .panel-icons{
    padding-top: 10px;
  }
  
  .panel-buttons{
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
  
  .warning-text{
    display: flex;
    justify-content: center;
  }

  .text-task{
    margin-top: 50px;
  }

  .button-delete-done{
    margin-top: 10px;
  }
  .v-slide-group__prev{
    display: none;
    margin-left: -40px;
  }
  .v-tab{
    font-size: 0.7rem !important;
    max-width: 150px !important
  }
  </style>
  