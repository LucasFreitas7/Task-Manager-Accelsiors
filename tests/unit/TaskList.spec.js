import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import TasksList from '@/components/TasksList.vue'
import replaceAllInserter from 'string.prototype.replaceall';

replaceAllInserter.shim();

const localVue = createLocalVue()

localVue.use(Vuex)

const formatDateMock = jest.fn()
const handleClickOnFilterMock = jest.fn()
const emitTaskEditMock = jest.fn()
const deleteTaskMock = jest.fn()
const toggleChangeMock = jest.fn()
describe('TasksList.vue', () => {
  let getters
  let store

  beforeEach(() => {
    getters = {
      getAll: () => [1,2,3],
      doneAll: () => [1, 2, 3],
      undoneAll: () => [1, 2, 3]
    }
    store = new Vuex.Store({
        getters
    })
  })
  it('Snapshot test', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({formatDate: formatDateMock})
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('When click on  All, must call handleClickOnFilter function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({handleClickOnFilter: handleClickOnFilterMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.find('v-tab').trigger('click')
    expect(handleClickOnFilterMock).toHaveBeenCalled()
  })
  it('When click on  Done Tasks, must call handleClickOnFilter function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({handleClickOnFilter: handleClickOnFilterMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.findAll('v-tab').at(1).trigger('click')
    expect(handleClickOnFilterMock).toHaveBeenCalled()
  })
  it('When click on  Undone Tasks, must call handleClickOnFilter function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({handleClickOnFilter: handleClickOnFilterMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.findAll('v-tab').at(2).trigger('click')

    expect(handleClickOnFilterMock).toHaveBeenCalled()
  })
  it('When click on editButton, must call emitTaskEdit function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({emitTaskEdit: emitTaskEditMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.findAll('v-btn').at(1).trigger('click')
    expect(emitTaskEditMock).toHaveBeenCalled()
  })
  it('When click on deleteButton, must call deleteTask function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({deleteTask: deleteTaskMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.findAll('v-btn').at(2).trigger('click')
    expect(deleteTaskMock).toHaveBeenCalled()
  })
  it('When click on deleteButton, must call deleteTask function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({deleteTask: deleteTaskMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.findAll('v-btn').at(1).trigger('click')
    expect(deleteTaskMock).toHaveBeenCalled()
  })
  it('When click on checkbox, must call toggleChange function', () => {
    const wrapper = shallowMount(TasksList, { store, localVue })
    wrapper.setMethods({toggleChange: toggleChangeMock})
    wrapper.setMethods({formatDate: formatDateMock})
    wrapper.find('v-btn').trigger('click')
    expect(toggleChangeMock).toHaveBeenCalled()
  })
})
