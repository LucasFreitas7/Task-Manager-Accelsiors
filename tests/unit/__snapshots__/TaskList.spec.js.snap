// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksList.vue Snapshot test 1`] = `
<v-container>
  <v-tabs>
    <v-tab>All tasks</v-tab>
    <v-tab>Done tasks</v-tab>
    <v-tab>Undone Tasks</v-tab>
  </v-tabs>
  <div class="warning-text">
    <!---->
    <!---->
    <!---->
  </div>
  <v-expansion-panels focusable="">
    <v-expansion-panel>
      <v-expansion-panel-header></v-expansion-panel-header>
      <v-expansion-panel-content>
        <div class="panel-icons">
          <v-icon>schedule</v-icon> = hours
        </div>
        <div class="panel-icons">
          <v-icon>comment</v-icon> =
        </div>
        <div class="panel-icons">
          <v-icon>calendar_month</v-icon> =
        </div>
        <div class="panel-buttons">
          <div>
            <v-btn color="error" class="ma-2">
              Undone Task
            </v-btn>
          </div>
          <div>
            <v-btn fab="" dark="" small="" color="cyan" class="mx-2">
              <v-icon dark="">
                mdi-pencil
              </v-icon>
            </v-btn>
            <v-btn fab="" dark="" small="" color="red" class="mx-2">
              <v-icon dark="">
                delete
              </v-icon>
            </v-btn>
          </div>
        </div>
      </v-expansion-panel-content>
    </v-expansion-panel>
    <v-expansion-panel>
      <v-expansion-panel-header></v-expansion-panel-header>
      <v-expansion-panel-content>
        <div class="panel-icons">
          <v-icon>schedule</v-icon> = hours
        </div>
        <div class="panel-icons">
          <v-icon>comment</v-icon> =
        </div>
        <div class="panel-icons">
          <v-icon>calendar_month</v-icon> =
        </div>
        <div class="panel-buttons">
          <div>
            <v-btn color="error" class="ma-2">
              Undone Task
            </v-btn>
          </div>
          <div>
            <v-btn fab="" dark="" small="" color="cyan" class="mx-2">
              <v-icon dark="">
                mdi-pencil
              </v-icon>
            </v-btn>
            <v-btn fab="" dark="" small="" color="red" class="mx-2">
              <v-icon dark="">
                delete
              </v-icon>
            </v-btn>
          </div>
        </div>
      </v-expansion-panel-content>
    </v-expansion-panel>
    <v-expansion-panel>
      <v-expansion-panel-header></v-expansion-panel-header>
      <v-expansion-panel-content>
        <div class="panel-icons">
          <v-icon>schedule</v-icon> = hours
        </div>
        <div class="panel-icons">
          <v-icon>comment</v-icon> =
        </div>
        <div class="panel-icons">
          <v-icon>calendar_month</v-icon> =
        </div>
        <div class="panel-buttons">
          <div>
            <v-btn color="error" class="ma-2">
              Undone Task
            </v-btn>
          </div>
          <div>
            <v-btn fab="" dark="" small="" color="cyan" class="mx-2">
              <v-icon dark="">
                mdi-pencil
              </v-icon>
            </v-btn>
            <v-btn fab="" dark="" small="" color="red" class="mx-2">
              <v-icon dark="">
                delete
              </v-icon>
            </v-btn>
          </div>
        </div>
      </v-expansion-panel-content>
    </v-expansion-panel>
  </v-expansion-panels>
  <div class="button-delete-done">
    <!---->
  </div>
</v-container>
`;
